### PR TITLE
Context-dependent spatial analysis preparation

### PR DESCRIPTION
**Pull request by alex-l-kong**
**What is the purpose of this PR?**

Begins to address #207. To prepare our implementation of context-dependent randomization in our spatial analysis, we make some changes to clarify certain functionalities of compute_close_cell_num and add parameters/return values as well.

**How did you implement your changes**

Splitting up compute_pos_cell_labels into channel and cluster functions. Adding parameters for specifying cell types to facet context-dependent randomization on. Adding a return value which maps marker labels to their respective cell types because we'll need this to randomize properly in a context-dependent sitting.

**Remaining issues**

More of a clarification. The reason we cannot implement context-dependent spatial analysis yet is because compute_close_cell_num_random does not take into account faceted randomization by cell_label and their respective marker counts. Changing the functionality of the compute_close_cell_num_random would make it overly complicated, so we'll need a separate randomization function in a context-dependent setting. That will be addressed in the next PR.

### CODE REVIEW

**Comment by alex-l-kong** [comment]
On a side note, we can probably fix coverage errors by adding simple tests for spatial_analysis_utils to cover cases where errors are raised and files are saved. Probably for another PR, though.

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Nice start, some more simplification that we can do

File: ark/analysis/spatial_analysis.py [hunk 1]
```diff
@@ -141,6 +140,9 @@ def calculate_cluster_spatial_enrichment(all_data, dist_matrices_dict, included_
             column with the cell phenotype IDs. Default is 'FlowSOM_ID'
         cell_label_col (str):
             column with the cell labels. Default is 'cellLabelInImage'
+        context_labels (dict):
+            A list that contains which specific types of cells we want to consider.
```
Comment (ackagel) [inline ark/analysis/spatial_analysis.py:144]:
```suggestion
            A dict that contains which specific types of cells we want to consider.
```

**Comment by alex-l-kong** [comment]
Sorry guys, got a bit carried away with documentation here. Reason is because it turns out I had to configure some parameters to get readthedocs to rebuild the documentation after splitting up get_pos_cell_labels.

File: ark/utils/spatial_analysis_utils.py [hunk 2]
```diff
@@ -84,34 +76,49 @@ def get_pos_cell_labels(analysis_type, pheno=None, current_fov_neighborhood_data
     Returns:
         list:
             List of all the positive labels"""
+    if(
+        thresh is None
+        or current_fov_channel_data is None
+        or cell_labels is None or current_marker is None
+    ):
+        raise ValueError("Incorrect arguments passed for analysis type")
```
Comment (ackagel) [inline ark/utils/spatial_analysis_utils.py:84]: Do we really need all these `None` checks, or are you just using this to find where all the old calls of `get_pos_cell_labels` are?

File: ark/utils/spatial_analysis_utils.py [hunk 3]
```diff
@@ -84,34 +76,49 @@ def get_pos_cell_labels(analysis_type, pheno=None, current_fov_neighborhood_data
...
+            the current cell phenotype
+        current_fov_neighborhood_data (pandas.DataFrame):
+            data for the current patient
+
+    Returns:
+        list:
+            List of all the positive labels"""
+
+    if pheno is None or current_fov_neighborhood_data is None:
+        raise ValueError("Incorrect arguments passed for analysis type")
```
Comment (ackagel) [inline ark/utils/spatial_analysis_utils.py:109]: see comment above

File: ark/utils/spatial_analysis_utils.py [hunk 4]
```diff
@@ -84,34 +76,49 @@ def get_pos_cell_labels(analysis_type, pheno=None, current_fov_neighborhood_data
...
-        ):
-            raise ValueError("Incorrect arguments passed for analysis type")
-        # Subset only cells that are positive for the given marker
-        marker1posinds = current_fov_channel_data[current_marker] > thresh
-        # Get the cell labels of the positive cells
-        mark1poslabels = cell_labels[marker1posinds]
     return mark1poslabels
 
 
 def compute_close_cell_num(dist_mat, dist_lim, num, analysis_type,
```
Comment (ackagel) [inline ark/utils/spatial_analysis_utils.py:119]: this was supposed to be fixed awhile ago, but got 'unfixed' from Jay and I's merge conflict.  I think `num` can just be computed from `cluster_ids` or `current_fov_channel_data`, and doesn't need to be passed in.

File: docs/Makefile [hunk 5]
```diff
@@ -1,14 +1,15 @@
...
 SOURCEDIR     = .
 BUILDDIR      = _build
 
-# Put it first so that "make" without argument is like "make help".
+SPHINXAPIDOC  = sphinx-apidoc
+MISCDOCFLAGS  = -f -E -T
+DOCOUTPUT     = md
+SRCPROJECT    = ../ark
+OUTPUTDIR     = _markdown
+IGNORE        = ../ark/analysis/*_test.py ../ark/segmentation/*_test.py ../ark/utils/*_test.py
```
Comment (ackagel) [inline docs/Makefile:11]: can't this just be `../ark/*/*_test.py` ?

**Review by ackagel — COMMENTED** [review]
Looks pretty super overall.  Just a few almost peripheral comments

File: ark/utils/spatial_analysis_utils.py [hunk 6]
```diff
@@ -58,20 +58,12 @@ def calc_dist_matrix(label_maps, path=None):
         np.savez(path + "dist_matrices.npz", **dist_matrices)
 
 
-def get_pos_cell_labels(analysis_type, pheno=None, current_fov_neighborhood_data=None,
-                        thresh=None, current_fov_channel_data=None, cell_labels=None,
-                        current_marker=None):
-    """Based on the type of the analysis, the function finds positive labels that match the
-    current phenotype or identifies cells with positive expression values for the current marker
+def get_pos_cell_labels_channel(thresh, current_fov_channel_data, cell_labels, current_marker):
```
Comment (ngreenwald) [inline ark/utils/spatial_analysis_utils.py:61]: Lets add a simple test function for both of these

File: ark/utils/spatial_analysis_utils_test.py [hunk 7]
```diff
@@ -143,15 +139,13 @@ def test_compute_close_cell_num():
...
-    channel_titles = fov_channel_data.columns
-    # Length of marker list
-    channel_num = len(channel_titles)
+
     # Subsetting threshold matrix to only include column with threshold values
     thresh_vec = example_thresholds.iloc[0:20, 1]
 
-    example_closenum, m1 = spatial_analysis_utils.compute_close_cell_num(
-        dist_mat=example_dist_mat, dist_lim=100, num=channel_num, analysis_type="channel",
+    # TODO: shouldn't we have a test somewhere for cluster analysis type?
```
Comment (ngreenwald) [inline ark/utils/spatial_analysis_utils_test.py:146]: Yes. I think we already have example data that should work for this in the test function for calc_cluster_spatial_enrichment. If not, even just a simple test that runs the function with the right inputs and doesn't check for correctness would be good to add.

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Just a few (hopefully) simple test functions to add

**Review by ackagel — COMMENTED** [review]
Just one thing from me.  I don't see `ark.utils.test_utils` in the docs.  Is it getting ignored from the `IGNORE` variable in the makefile, or is that default sphinx behavior to ignore `test_*` files?

**Comment by alex-l-kong** [comment]
> Just one thing from me. I don't see `ark.utils.test_utils` in the docs. Is it getting ignored from the `IGNORE` variable in the makefile, or is that default sphinx behavior to ignore `test_*` files?

It's because I had to resolve a merge conflict and I accidentally deleted that submodule. I'll re-add it.

**Review by ngreenwald — APPROVED** [review]
Nice stuff

**Pull request merged**